### PR TITLE
Add a StringList type to be used for commandline flags.

### PR DESCRIPTION
**Pull request by nilayvaish**
StringList flags would used for developing the feature: add tags as
pseudo locations.
#558

### CODE REVIEW

File: internal/stringlist/stringlist.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// Package stringlist provides type StringList that can be used as a type for
+// commandline flags for providing a slice of strings as input.
+package stringlist
+
+import (
+	"fmt"
+)
+
+// StringList is an alias for slice of strings.
+type StringList []string
```
Comment (nolanmar511) [inline internal/stringlist/stringlist.go:24]: Could []string be used directly?

I'm not immediately seeing the additional utility of having a wrapper class here, though could be missing something.

File: driver/driver.go [hunk 2]
```diff
@@ -94,7 +95,7 @@ type FlagSet interface {
 
 	// StringList is similar to String but allows multiple values for a
 	// single flag
-	StringList(name string, def string, usage string) *[]*string
+	StringList(name string, def string, usage string) *stringlist.StringList
```
Comment (nolanmar511) [inline driver/driver.go:98]: This is part of pprof's public interface (one passes in `plugin.Options` to `driver.PProf`; we use this internally to, among other things, change the profile fetcher / symbolizer for internal pprof). 

It would be nice if this interface did not have to change. Since we use this interface internally, it will definitely make this PR simpler to land if there aren't changes to pprof's public interface.

**Review by nolanmar511 — COMMENTED** [review]
Thanks for this change!

Left a few initial comments about the StringList type.

File: driver/driver.go (hunk 2 above)
Comment (aalexand) [reply driver/driver.go:98]: Also, this public interface shouldn't use an internal type (`github.com/google/pprof/internal/stringlist`) in its signature.

Comment (nilayvaish) [reply driver/driver.go]: I think the interface needs to change.  Currently there is no difference between a StringList flag and a String flag.  A StringList flag can only hold a single string for now.  To support a list of strings, we need to provide a type that adds strings to a list.  The new type implements the [flag.Value interface](https://golang.org/src/flag/flag.go?s=32508:32557#L291) to achieve this.

> Also, this public interface shouldn't use an internal type (github.com/google/pprof/internal/stringlist) in its signature.

I have moved the new type to a public directory.

File: internal/stringlist/stringlist.go (hunk 1 above)
Comment (nilayvaish) [reply internal/stringlist/stringlist.go]: The new type implements [the flag.Value interface](https://golang.org/src/flag/flag.go?s=32508:32557#L291).  I think we cannot retain the []string type and provide the functions required by the flag package.  I am open to considering other ways to achieve this, but there does not seem to be any.

File: driver/driver.go (hunk 2 above)
Comment (nolanmar511) [reply driver/driver.go]: I think I'd need to see a design document considering this more carefully proposed change before I'd be comfortable changing the public interface. I'm not currently persuaded that the interface change is the best or only way forward; I'd really need to be convinced of that before I'd feel comfortable with a change to the interface. 

As additional context, changing the interface is a really big deal. First, we don't currently version pprof. Now that golang does support versioning a library, if we change the public interface, this becomes something we ought to look into. IMO, it'd be necessary to at least fully consider versioning pprof  using go modules prior to making any breaking changes to the public interface. Next, changes to the public interface can make it really hard to sync our internal copy of pprof, since it becomes possible that code used for our internal version of pprof will break. If it is concluded that changes to the public interface are necessary, I would prefer to first see changes made to our internal copy of pprof to know that nothing internal will break (a PR would still need to be sent to change code, and any changes that need to be made to our dependencies on this library would need to be done when importing pprof from GitHub).

File: flagset/flagset.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2021 Google Inc. All Rights Reserved.
```
Comment (aalexand) [inline flagset/flagset.go:1]: I like pulling this interface out to avoid duplication between driver/driver.go and internal/plugin/plugin.go. Can we make this a separate PR? I think this can go in regardless of other discussions. One thing I'm not sure about is whether it's worth a separate top-level directory - seems too shallow for that. Maybe have it as driver/flagset.go? (and "package driver").

Comment (nilayvaish) [reply flagset/flagset.go:1]: > Can we make this a separate PR?

Yes.  We just need to decide on the package for flagset.go.

> Maybe have it as driver/flagset.go?

I had tried adding to driver package before, but ran into an import cycle.
package github.com/google/pprof
	imports github.com/google/pprof/driver
	imports github.com/google/pprof/internal/driver
	imports github.com/google/pprof/internal/binutils
	imports github.com/google/pprof/internal/plugin
	imports github.com/google/pprof/driver: import cycle not allowed
package github.com/google/pprof
	imports github.com/google/pprof/driver
	imports github.com/google/pprof/internal/driver
	imports github.com/google/pprof/internal/binutils
	imports github.com/google/pprof/internal/plugin
	imports github.com/google/pprof/driver: import cycle not allowed

Are you OK going with a new package?  I could not think of any other reasonable place for this.

Comment (aalexand) [reply flagset/flagset.go:1]: I'd prefer not to have a new top-level package. Actually, reading go/go-style/decisions#interfaces I wonder if we should leave things as is with two copies of the interface for now. It's duplication, but it's rather unrelated issue to this PR, so I'd tackle it separately if we ever need to.

File: driver/driver.go (hunk 2 above)
Comment (nilayvaish) [reply driver/driver.go]: I have updated the PR with a different implementation (suggested by lmilewski) that does not change the public interface.  Please take another look.

File: internal/driver/flags.go [hunk 4]
```diff
@@ -44,9 +45,33 @@ func (*GoFlags) String(o, d, c string) *string {
...
+func (sl *stringList) Set(v string) error {
+	*sl = append(*sl, &v)
+	return nil
+}
+
+func (sl *stringList) Get() interface{} {
+	return sl
+}
+
+func (sl *stringList) String() string {
```
Comment (nolanmar511) [inline internal/driver/flags.go:59]: For my understanding, where (other than tests) is this method used?

File: internal/driver/flags_test.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package driver
+
+import (
+	"testing"
+)
+
+func TestString(t *testing.T) {
```
Comment (nolanmar511) [inline internal/driver/flags_test.go:21]: It seems a bit different (at least to the best of my knowledge and in the context of golang) to test functions on a private type in golang. Testing the public interface would probably be preferable.

But, I know that there are no tests for flag.go right now; will leave it up to @aalexand  to decide what sort of testing is sufficient.

File: internal/driver/flags.go (hunk 4 above)
Comment (nilayvaish) [reply internal/driver/flags.go:59]: This function is required to implement the flag.Value interface.  Link: https://golang.org/pkg/flag/#Value.

File: internal/driver/flags_test.go (hunk 5 above)
Comment (nolanmar511) [reply internal/driver/flags_test.go:21]: Ping to @nilayvaish and @aalexand.

I'm not sure about this testing. It may be better than not having tests, but I'm a bit skeptical about testing only a private type, rather than the publicly available functions.

Comment (nilayvaish) [reply internal/driver/flags_test.go:21]: AFAICT, it is not possible to test the public interface without creating a separate process.  Would you be OK with the _test.go file creating a new process with flags?  Otherwise, I think we need to drop the tests all together.

**Comment by aalexand** [comment]
@nilayvaish @nolanmar511 Looking at this again, I'm now not sure why flag changes are needed to support `-tagroot` / `-tagleaf`. Those shouldn't be flags, should be vars in internal/driver/config.go like `-tagfocus` / `-tagignore` and friends.

Also, one thing to watch for with list-like flags is append vs. reset semantics. E.g. in the pprof interactive shell we need a way to specify the full set of values for these flags, without ad-infinitum append behavior.

**Comment by nilayvaish** [comment]
> @nilayvaish @nolanmar511 Looking at this again, I'm now not sure why flag changes are needed to support `-tagroot` / `-tagleaf`. I believe those shouldn't be flags, I think they should be vars in internal/driver/config.go like `-tagfocus` / `-tagignore` and friends.

We are treating `tagroot` and `tagleaf` like `tagfocus and `tagignore`, only the types are different.  I have added another commit to show this.

> 
> Also, one thing to watch for with list-like flags is append vs. reset semantics. E.g. in the pprof interactive shell we need a way to specify the full set of values for these flags, without ad-infinitum append behavior.

This is an interesting point I had not considered before.  I checked [`internal/drive/interactive.go`](https://github.com/google/pprof/blob/master/internal/driver/interactive.go) and it seems like the command line is parsed again.  `TagIgnore`and `TagFocus` have been treated specially in the parse routing: [`parseCommandLine`](https://github.com/google/pprof/blob/master/internal/driver/interactive.go#L219).  We could do something similar for `TagRoot` and `TagLeaf`.  Let me know what you think.

File: internal/driver/commands.go [hunk 6]
```diff
@@ -199,6 +199,14 @@ var configHelp = map[string]string{
...
 		"Numeric tag filter examples: 1kb, 1kb:10kb, memory=32mb:",
 		"String tag filter examples: foo, foo.*bar, mytag=foo.*bar"),
+	"tagleaf": helpText(
+		"Adds a pseudo leaf node for samples with tags matched by regexp",
+		"Use name=value1,value2 syntax to specify the tag and the values",
+		"for which nodes would be added.  Example: latency=1s,2s,3s."),
+	"tagroot": helpText(
+		"Adds a pseudo root node for samples with tags matched by regexp",
+		"Use name=value1,value2 syntax to specify the tag and the values",
+		"for which nodes would be added.  Example: latency=1s,2s,3s."),
```
Comment (aalexand) [inline internal/driver/commands.go:209]: Couple of questions:

1) The option / flag is of type []string / multi-value. The examples are of the form "foo=val1,val2" What exactly the multi-value here? Is it the comma-separated multi-values as in example, i.e. "foo=val1" and "val2" as two values which is weird? Or is that there is ability to have multiple occurrences of each of these like "-flag=foo=val1,val2 -flag=bar=val3,val4" and the documentation needs to explain this?

2) How are numeric tags handled, including with units like memory? Can we reuse the same value format as in tagfocus / taghide etc.? I feel that we are introducing a new way to specify a set of values and I don't see why different tag* flags can't unify on a single way to do that.
